### ISSUE TITLE
[feat] Support for podman

### ISSUE THREAD

**Issue by cardil** [knative-sandbox/kn-plugin-func#188]
**Is your feature request related to a problem? Please describe.**
On my Fedora box with podman installed I'm unable to use FaaS. I've been using podman happily for more than a year without any trouble. I can do all kinds of stuff with podman and I have never been forced to have docker installed. I would like to use faas without hasle and security pitfalls of having Docker daemon running.

FaaS relying on Docker socket is in conflict with a RH narration of ["no big fat daemons"](https://av.tib.eu/media/43121).

**Describe the solution you'd like**
FaaS should call out to OS installed OCI tool like podman. Things will even work if faas would just call `docker ...` process as there is a podman-docker package on Fedora.

Also, there is an option to vendor https://github.com/containers/image (this one is used by buildah, podman, and skopeo, and have most of the needed code) and use it to build images without any external tool being installed.

**Describe alternatives you've considered**
I need to run a VM with Docker installed to use faas, because of this. I can't install Docker and Podman on the same machine because of cgroups v2 incompatibility.

**Comment by matejvasek** [comment]
This depends on https://github.com/boson-project/func/issues/179

**Comment by matejvasek** [comment]
@cardil I've chosen path of using `podman` as a `docker` deamon: `podman system service`.
User can do:
```
$ podman system service --time=0 tcp:0.0.0.0:1234 & # let run in background or another terminal
$ export DOCKER_HOST=tcp://127.0.0.1:1234
$ func build -v
```

**Comment by cardil** [comment]
But, that's not how Podman is supposed to work. Podman was done to get rid of a daemon. `kn func` should support podman as it is, not by calling an unsecured public port. If you publish that service to `tcp:0.0.0.0:1234` anyone could connect to your machine and do anything you can do on it - that's quite a **BIG** vulnerability. `kn func` shouldn't require opening such vulnerability to operate on RHEL 8/Fedora machines.

A minimal thing would be to support unix sockets (default: `unix:/$XDG_RUNTIME_DIR/podman/podman.sock`) or better to just call podman as a library (libpod) or as child executable.

**Comment by matejvasek** [comment]
@cardil

That was bad example from me, you don't have to run `tcp:0.0.0.0:1234` it would work with `tcp:127.0.0.1:1234` too.

Also the Unix socket works, but you need to setup proper policy (or disable labeling for containers) when using SELinux.
We are looking into it; see https://github.com/containers/podman/discussions/9860.

Golang libpod binding library is not deamon less. It's just a REST client and it requires podman deamon running on the machine.

**Comment by matejvasek** [comment]
We are using buildpack's `pack` to build function; it uses Docker APIv2. Reimplementing `pack` would be a lot of work.

**Comment by matejvasek** [comment]
Also this approach made `pack` CLI usable with podman which is beneficial also people outside boson's `func`.

**Comment by cardil** [comment]
`tcp:127.0.0.1:1234` is as vulnerable on server environments as `tcp:0.0.0.0:1234`. Even on more standard single user machine `tcp:127.0.0.1:1234` opens up a way for malicious code to attack a user.

**Comment by matejvasek** [comment]
> `tcp:127.0.0.1:1234` is as vulnerable on server environments as `tcp:0.0.0.0:1234`. Even on more standard single user machine `tcp:127.0.0.1:1234` opens up a way for malicious code to attack a user.

@cardil then Unix socket it is. We just have to prepare SELinux policies or disable labeling of build containers.

**Comment by matejvasek** [comment]
@cardil to explain a little the issue with SELinux:
The `pack` runs build in a container, the container needs a way to push resulting image to local repository. That's why it mount's the socket. By default it's forbidden since mounting socket is potential security risk (especially if daemon runs as the root).

There is one more way around mounting: not to push to local repository, but push to remote container registry. One disadvantage is that you need to push it to remote and then pull it back.

**Comment by lance** [comment]
@matejvasek I think this issue has been resolved, correct? Can it be closed now?

**Comment by matejvasek** [comment]
as of `podman` 3+ is should work.

**Comment by cardil** [comment]
I got podman 3.1.2, and  it still has some problems:

```
$ systemctl status --user podman.socket
● podman.socket - Podman API Socket
     Loaded: loaded (/usr/lib/systemd/user/podman.socket; disabled; vendor preset: disabled)
     Active: active (listening) since Tue 2021-06-15 21:14:35 CEST; 16s ago
   Triggers: ● podman.service
       Docs: man:podman-system-service(1)
     Listen: /run/user/1000/podman/podman.sock (Stream)
     CGroup: /user.slice/user-1000.slice/user@1000.service/podman.socket

Jun 15 21:14:35 thinkpad-t590 systemd[3381]: Listening on Podman API Socket.

$ export DOCKER_HOST=unix://${XDG_RUNTIME_DIR}/podman/podman.sock
$ kn func build --verbose
A registry for function images is required (e.g. 'quay.io/boson').

Registry for function images: quay.io/cardil
Building function image
2021/06/15 21:15:22.842538 DEBUG:  Pulling image quay.io/boson/faas-quarkus-jvm-builder:latest
cb0ba8718831: Already exists 
33db084abe90: Already exists 
d0c9851d609d: Already exists 
00b73a81ae14: Already exists 
35dfa5837bb4: Already exists 
0034ab586d6b: Already exists 
4f4fb700ef54: Already exists 
d17e12a6c002: Download complete 
52608334c58a: Download complete 
c96129f1a707: Download complete 
de301fec2d36: Download complete 
8e5aaa35e8d3: Pull complete 
2021/06/15 21:15:36.477690 DEBUG:  Selected run image quay.io/boson/faas-stack-run:quarkus-jvm-v0.8.1
2021/06/15 21:15:37.943482 DEBUG:  Pulling image quay.io/boson/faas-stack-run:quarkus-jvm-v0.8.1
159e7c5ccea1: Download complete 
6eb9436aed48: Download complete 
b68d652c1331: Download complete 
1efc9d5eb7cf: Download complete 
1b4ac40997ba: Download complete 
59925e427b6c: Download complete 
447984ba0244: Pull complete 
2021/06/15 21:15:51.257305 DEBUG:  Creating builder with the following buildpacks:
2021/06/15 21:15:51.257320 DEBUG:  -> dev.boson.quarkus-jvm@0.0.1
2021/06/15 21:16:02.821124 DEBUG:  Pulling image buildpacksio/lifecycle:0.10.1
f255694a0a1f: Already exists 
e59bd8947ac7: Already exists 
3066741dcf79: Pull complete 
2021/06/15 21:16:07.828170 DEBUG:  Using build cache volume pack-cache-cardil_translate_latest-1d36dcac0a35.build
2021/06/15 21:16:07.828186 INFO:   ===> DETECTING
2021/06/15 21:16:07.828206 DEBUG:  Running the detector on OS linux with:
2021/06/15 21:16:07.828210 DEBUG:  Container Settings:
2021/06/15 21:16:07.828213 DEBUG:    Args: /cnb/lifecycle/detector
2021/06/15 21:16:07.828215 DEBUG:    System Envs: CNB_PLATFORM_API=0.4
2021/06/15 21:16:07.828219 DEBUG:    Image: pack.local/builder/656f6a6b786f6e737a7a:latest
2021/06/15 21:16:07.828223 DEBUG:    User: 
2021/06/15 21:16:07.828231 DEBUG:    Labels: map[author:pack]
2021/06/15 21:16:07.828233 DEBUG:  Host Settings:
2021/06/15 21:16:07.828236 DEBUG:    Binds: pack-layers-lkeoyualll:/layers pack-app-fgaivujgvx:/workspace
2021/06/15 21:16:07.828240 DEBUG:    Network Mode: host
[detector] dev.boson.quarkus-jvm 0.0.1
2021/06/15 21:16:08.509557 INFO:   ===> ANALYZING
2021/06/15 21:16:08.509596 DEBUG:  Running the analyzer on OS linux with:
2021/06/15 21:16:08.509602 DEBUG:  Container Settings:
2021/06/15 21:16:08.509629 DEBUG:    Args: /cnb/lifecycle/analyzer -daemon -cache-dir /cache quay.io/cardil/translate:latest
2021/06/15 21:16:08.509636 DEBUG:    System Envs: CNB_USER_ID=1000 CNB_GROUP_ID=1001 CNB_PLATFORM_API=0.4
2021/06/15 21:16:08.509641 DEBUG:    Image: buildpacksio/lifecycle:0.10.1
2021/06/15 21:16:08.509645 DEBUG:    User: root
2021/06/15 21:16:08.509657 DEBUG:    Labels: map[author:pack]
2021/06/15 21:16:08.509660 DEBUG:  Host Settings:
2021/06/15 21:16:08.509666 DEBUG:    Binds: /run/user/1000/podman/podman.sock:/var/run/docker.sock pack-cache-cardil_translate_latest-1d36dcac0a35.build:/cache pack-layers-lkeoyualll:/layers pack-app-fgaivujgvx:/workspace
2021/06/15 21:16:08.509674 DEBUG:    Network Mode: host
[analyzer] ERROR: failed to get previous image: Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Get "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/info": dial unix /var/run/docker.sock: connect: permission denied
Error: executing lifecycle. This may be the result of using an untrusted builder: failed with status code: 1
```

**Comment by cardil** [comment]
@matejvasek As Dan [pointed out](https://github.com/containers/podman/discussions/9860#discussioncomment-545142), adding `--security-opt label=disabled` should help this, right? 

But, that needs to be added via REST, not as a CLI option. Am I right?

**Comment by cardil** [comment]
Podman extends the Docker API a bit. So, I think it might be possible to detect if given endpoint is podman 3+ and use this API to add that security-opt to analyzer container: https://docs.podman.io/en/latest/_static/api.html#operation/ContainerCreateLibpod

**Comment by cardil** [comment]
Although, the code below uses build pack client:

https://github.com/boson-project/func/blob/a937c490b7e1ad31c3596f91c310c3f4560329fd/buildpacks/builder.go#L65-L91

That means a proper handling of this `label=disabled` security opt should land on https://github.com/buildpacks/pack project first...

**Comment by cardil** [comment]
Lol. I've just dig out your commit there, that added the `DockerHost` opt: https://github.com/buildpacks/pack/commit/be9b1b9c6f7d8cffff18b3dd9a9c7b7e9caa173f

So, my proposal. Extend https://github.com/buildpacks/pack once more, so that it will properly handle the podman socket, by adding `label=disabled` security opt to containers that mount the docker socket. I'll create such issue in https://github.com/buildpacks/pack project.

**Comment by matejvasek** [comment]
> @matejvasek As Dan pointed out, adding --security-opt label=disabled should help this, right?

Yeah if I recall correctly it does help.

**Comment by matejvasek** [comment]
`--security-opt` is valid `docker` flag (it's not `podman` specific) maybe we could add the flag for `pack`.
Another solution/workaround would be use `containers.conf` where you can disable labeling for all `podman` containers.

**Comment by matejvasek** [comment]
> `--security-opt` is valid `docker` flag (it's not `podman` specific) maybe we could add the flag for `pack`.
> Another solution/workaround would be use `containers.conf` where you can disable labeling for all `podman` containers.

since it's standard docker flag I think there is a change of getting it into `pack`; if it were just `podman` specific I think maintainers might not like it

**Comment by matejvasek** [comment]
@cardil but you will probably encounter the error also with `moby` not just `podman`.

**Comment by matejvasek** [comment]
https://github.com/buildpacks/pack/issues/1208

**Comment by matejvasek** [comment]
@cardil  What be could do now (before the functionality will be added to the `pack`):

```diff
diff --git a/buildpacks/builder.go b/buildpacks/builder.go
index fc3f766..ece0e76 100644
--- a/buildpacks/builder.go
+++ b/buildpacks/builder.go
@@ -82,7 +82,7 @@ func (builder *Builder) Build(ctx context.Context, f bosonFunc.Function) (err er
        }
 
        // Client with a logger which is enabled if in Verbose mode.
-       packClient, err := pack.NewClient(pack.WithLogger(logging.New(logWriter)))
+       packClient, err := pack.NewClient(pack.WithLogger(logging.New(logWriter)), pack.WithDockerClient(wrapperwithSecOptInjected))
        if err != nil {
                return
        }
```

It's possible to set custom docker client for the `pack` client.
We could create custom wrapper for standard docker client, the wrapper would just inject the flag sec-opts flag for `ContainerCreate()` function.

**Comment by cardil** [comment]
That would be awesome if it works ☺️

**Comment by cardil** [comment]
Shouldn't this be closed because #390 is merged?

**Comment by matejvasek** [comment]
for how to use `podman` see https://github.com/knative-sandbox/kn-plugin-func/blob/main/docs/guides/podman.md
